nixos/modules/services/hardware/display.nix

```diff
@@ -0,0 +1,141 @@
+        and exposed as `edid/<name>.bin` files in initrd.
+        See for more information:
+        - `hardware.display.edid.packages`
+        - https://en.wikipedia.org/wiki/XFree86_Modeline
+      '';
+      example = lib.literalExpression ''
+        {
+          "PG278Q_60" = "    241.50   2560 2608 2640 2720   1440 1443 1448 1481   -hsync +vsync";
+          "PG278Q_120" = "   497.75   2560 2608 2640 2720   1440 1443 1448 1525   +hsync -vsync";
+          "U2711_60" = "     241.50   2560 2600 2632 2720   1440 1443 1448 1481   -hsync +vsync";
+        }
+      '';
+      apply = modelines:
+        if modelines == { } then null else
+        pkgs.edid-generator.overrideAttrs {
+          clean = true;
+          modelines = lib.trivial.pipe modelines [
+            (lib.mapAttrsToList (name: value:
+              lib.throwIfNot (builtins.stringLength name <= 12) "Modeline name must be 12 characters or less"
+                ''Modeline "${name}" ${value}''
+            ))
+            (builtins.map (line: "${line}\n"))
+            (lib.strings.concatStringsSep "")
+          ];
+        };
+    };
+
+    hardware.display.outputs = lib.mkOption {
+      type = lib.types.attrsOf (lib.types.submodule ({
+        options = {
+          edid = lib.mkOption {
+            type = with lib.types; nullOr str;
+            default = null;
+            description = lib.mdDoc ''
+              An EDID filename to be used for configured display, as in `edid/<filename>`.
+              See for more information:
+              - `hardware.display.edid.packages`
+              - https://wiki.archlinux.org/title/Kernel_mode_setting#Forcing_modes_and_EDID
+            '';
+          };
+          mode = lib.mkOption {
+            type = with lib.types; nullOr str;
+            default = null;
+            description = lib.mdDoc ''
+              An `video` kernel parameter (framebuffer mode) configuration for the specific output:
+
+                 <xres>x<yres>[M][R][-<bpp>][@<refresh>][i][m][eDd]
+
+              See for more information:
+              - https://docs.kernel.org/fb/modedb.html
+              - https://wiki.archlinux.org/title/Kernel_mode_setting#Forcing_modes
+            '';
+            example = lib.literalExpression ''
+              "e"
+            '';
+          };
+        };
+      }));
+      description = lib.mdDoc ''
+        Hardware/kernel-level configuration of specific outputs.
+      '';
+      default = { };
+
+      example = lib.literalExpression ''
+        {
+          edid.modelines."PG278Q_60" = "241.50   2560 2608 2640 2720   1440 1443 1448 1481   -hsync +vsync";
+          outputs."DP-1".edid = "PG278Q_60.bin";
+          outputs."DP-1".mode = "e";
+        }
+      '';
+    };
+  };
+
+  ###### implementation
+
+  # services.udev implements hardware.firmware option,
+  # there is no point running EDIDs without it
+  config = lib.mkIf config.services.udev.enable {
```

Comment:
Maybe it is cleaner to enabled udev below and depend on cfg.edid.packages here?

Reply:
what do you think about the new version?